### PR TITLE
notepad-next: init at 0.5.1

### PR DESCRIPTION
**Pull request by SebTM**
###### Description of changes

Add NotepadNext at v0.5

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SebTM** [comment]
Can I do something more to get this merged?

File: pkgs/applications/editors/notepad-next/default.nix [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  # References
+  #  https://github.com/dail8859/NotepadNext/blob/master/doc/Building.md
+  #  https://github.com/dail8859/NotepadNext/issues/99
+  buildPhase = ''
+    mkdir -p $out/bin $out/share/applications
+    substituteInPlace ./src/NotepadNext/NotepadNext.pro --replace /usr $out
+    mkdir build
+    cd build
+    qmake ../src/NotepadNext.pro
```
Comment (Ma27) [inline pkgs/applications/editors/notepad-next/default.nix:23]: This can be done via
``` nix
{
  nativeBuildInputs = [ qmake ];
  qmakeTargets = [ "src/NotepadNext.pro" ];
}
```

If you also move the `substituteInPlace` into `postPatch`, you don't even need a custom `buildPhase` :)

**Review by Ma27 — COMMENTED** [review]
One minor remark. Also, please rename the commi tmessage to `notepad-next: init at 0.5` to match your contribution guidelines.

**Comment by SebTM** [comment]
Thanks for your help and feedback - I've updated the PR and changed to the latest version :)

File: pkgs/applications/editors/notepad-next/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  };
+
+  nativeBuildInputs = [ qmake libsForQt5.qt5.qttools ];
+  qmakeFlags = [ "src/NotepadNext.pro" ];
+
+  # References
+  #  https://github.com/dail8859/NotepadNext/blob/master/doc/Building.md
+  #  https://github.com/dail8859/NotepadNext/issues/99
+  postPatch = ''
+    substituteInPlace ./src/NotepadNext/NotepadNext.pro --replace /usr $out
```
Comment (jtojnar) [inline pkgs/applications/editors/notepad-next/default.nix:22]: Do you meant to link https://github.com/dail8859/NotepadNext/pull/124?

File: pkgs/applications/editors/notepad-next/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  '';
+
+  CXXFLAGS = "-std=gnu++1z";
+
+  enableParallelBuilding = true;
```
Comment (jtojnar) [inline pkgs/applications/editors/notepad-next/default.nix:27]: `enableParallelBuilding` is enabled by default in qmake packages. See <https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unnecessary-parallel-building.md> for more info.

File: pkgs/applications/editors/notepad-next/default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  meta = with lib; {
+    homepage = "https://github.com/dail8859/NotepadNext";
+    description = "Notepad++-like editor for the Linux desktop";
+    license = licenses.gpl3;
```
Comment (jtojnar) [inline pkgs/applications/editors/notepad-next/default.nix:32]: `gpl3` is unclear, use `gpl3Plus` or `gpl3Only`. See https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

File: pkgs/applications/editors/notepad-next/default.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  qmakeFlags = [ "src/NotepadNext.pro" ];
+
+  # References
+  #  https://github.com/dail8859/NotepadNext/blob/master/doc/Building.md
+  #  https://github.com/dail8859/NotepadNext/issues/99
+  postPatch = ''
+    substituteInPlace ./src/NotepadNext/NotepadNext.pro --replace /usr $out
+  '';
+
+  CXXFLAGS = "-std=gnu++1z";
```
Comment (jtojnar) [inline pkgs/applications/editors/notepad-next/default.nix:25]: Why is this necessary? Could not upstream set appropriate C++ requirements in their build scripts?

File: pkgs/applications/editors/notepad-next/default.nix [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+mkDerivation rec {
+  pname = "notepad-next";
+  version = "0.5.1";
+
+  src = fetchFromGitHub {
+    owner = "dail8859";
+    repo = "NotepadNext";
+    rev = "v${version}";
+    sha256 = "sha256-J7Ngt6YtAAZsza2lN0d1lX3T8gNJHp60sCwwaLMGBHQ=";
+    fetchSubmodules = true;
```
Comment (jtojnar) [inline pkgs/applications/editors/notepad-next/default.nix:12]: Would be even nicer if upstream optionally supported external packages instead of vendored dependencies. We already have lua packaged and it should not be that hard to package the rest.

File: pkgs/applications/editors/notepad-next/default.nix (hunk 2 above)
Comment (SebTM) [reply pkgs/applications/editors/notepad-next/default.nix:22]: I refereed more to https://github.com/dail8859/NotepadNext/issues/99#issuecomment-1100761965 but I can change it to this as well as mentioned in https://github.com/dail8859/NotepadNext/issues/126#issuecomment-1120403377 we should not need it in the future when a release with this change is available - even I don't already know how to use the new prefix option.

File: pkgs/applications/editors/notepad-next/default.nix (hunk 5 above)
Comment (SebTM) [reply pkgs/applications/editors/notepad-next/default.nix:25]: I got this hint here: https://github.com/dail8859/NotepadNext/issues/99#issuecomment-1100762910
It decreases the amount of warnings while building...

File: pkgs/applications/editors/notepad-next/default.nix (hunk 6 above)
Comment (SebTM) [reply pkgs/applications/editors/notepad-next/default.nix:12]: How can I verify if this is possible/is there a reference that I can suggest upstream how to do it? Maybe we can make an issue out of this in the end not blocking a merge?

File: pkgs/applications/editors/notepad-next/default.nix (hunk 2 above)
Comment (jtojnar) [reply pkgs/applications/editors/notepad-next/default.nix:22]: The qmake setup hook should set the prefix by default: https://github.com/NixOS/nixpkgs/blob/634141959076a8ab69ca2cca0f266852256d79ee/pkgs/development/libraries/qt-5/hooks/qmake-hook.sh#L10

File: pkgs/applications/editors/notepad-next/default.nix (hunk 6 above)
Comment (jtojnar) [reply pkgs/applications/editors/notepad-next/default.nix:12]: Yeah, this will be a bit more work upstream so issue should be fine. I am not really familiar with qmake but generally, there would be an option to enable using external dependencies ([example for CMake based project](https://github.com/NixOS/nixpkgs/pull/171934/files#diff-afc2a5b63cab6314c97002777c79efe92ad10559b2c482b95ecfff0aa574bf5eR161)), which when enabled, would find the dependencies using pkg-config instead of building them locally.

For example, here are Debian guidelines for projects concerning this: https://wiki.debian.org/UpstreamGuide#No_inclusion_of_third_party_code

File: pkgs/applications/editors/notepad-next/default.nix (hunk 5 above)
Comment (jtojnar) [reply pkgs/applications/editors/notepad-next/default.nix:25]: They should probably change https://github.com/dail8859/NotepadNext/blob/8bcfc93ba41f8bc1150385864f8adf501d842762/src/Config.pri#L24 to what C++ standards they expect, then.

Comment (SebTM) [reply pkgs/applications/editors/notepad-next/default.nix:27]: Thanks reported it upstream and keep track of it in the follow releases as package-maintainer :)

File: pkgs/applications/editors/notepad-next/default.nix (hunk 6 above)
Comment (SebTM) [reply pkgs/applications/editors/notepad-next/default.nix:13]: Thanks for the help, I reported it upstream with reference to your comment and referenced the issue in the code for future updates/changes 👍🏻  https://github.com/dail8859/NotepadNext/issues/135

**Comment by SebTM** [comment]
https://github.com/dail8859/NotepadNext/issues/135#issuecomment-1121765043

Maybe @jtojnar or someone else can help me answering Maintainers question about Flatpack-Package provided, and a more comprehensive explanation about our request to make external dependencies possible as for now there are no desires to change it.

**Comment by jtojnar** [comment]
Some people prefer native packages over flatpak for various reasons (simplicity of native packaging, (especially space) overhead of Flatpak, current inability to install software declaratively on Flatpak, trivial local patching…).

But if CMake port is planned, it makes sense to wait for it since it is much easier to achieve unbundling – CMake’s [`find_package`](https://cmake.org/cmake/help/latest/command/find_package.html) function is pretty good.

**Comment by SebTM** [comment]
@jojnar Top :) Thanks for helping out - are we ready to merge this until a new release is out? I will remove the current fixes then 👍🏻

**Comment by SebTM** [comment]
> @jojnar Top :) Thanks for helping out - are we ready to merge this until a new release is out? I will remove the current fixes then 👍🏻

@jtojnar sorry, wrong mention :)

**Review by Ma27 — APPROVED** [review]

**Comment by Mindavi** [comment]
Broken on aarch64, consider marking broken or explictly marking supported platforms: https://hydra.nixos.org/build/177847891/nixlog/2

**Comment by SebTM** [comment]
> Broken on aarch64, consider marking broken or explictly marking supported platforms: https://hydra.nixos.org/build/177847891/nixlog/2

https://github.com/NixOS/nixpkgs/pull/174062/files - done 👍🏻
